src/NuGet.Core/NuGet.Commands/RestoreCommand/VulnerabilityInformationProvider.cs

```diff
@@ -13,38 +13,53 @@
 
 namespace NuGet.Commands
 {
-    internal sealed class VulnerabilityInformationProvider : IVulnerabilityInformationProvider
+    internal sealed class VulnerabilityInformationProvider : IVulnerabilityInformationProvider, IDisposable
     {
+        private readonly SemaphoreSlim _semaphore = new(initialCount: 1, maxCount: 1);
         private readonly SourceRepository _source;
         private readonly ILogger _logger;
 
-        private readonly AsyncLazy<GetVulnerabilityInfoResult?> _vulnerabilityInfo;
+        private GetVulnerabilityInfoResult? _cachedResult;
 
         public VulnerabilityInformationProvider(SourceRepository source, ILogger logger, bool isAuditSource)
         {
             _source = source;
             _logger = logger;
             IsAuditSource = isAuditSource;
-
-            _vulnerabilityInfo = new AsyncLazy<GetVulnerabilityInfoResult?>(GetVulnerabilityInfoAsync);
         }
 
         public async Task<GetVulnerabilityInfoResult?> GetVulnerabilityInformationAsync(CancellationToken cancellationToken)
         {
-            GetVulnerabilityInfoResult? result = await _vulnerabilityInfo;
-            return result;
+            await _semaphore.WaitAsync(cancellationToken);
+
+            try
+            {
+                if (_cachedResult is not null)
+                {
+                    return _cachedResult;
+                }
```

Comment:
You could duplicate this block before the semaphore wait, so that once it's cached, there's no longer work needed to get and release the semaphore.  In a large solution restore, this early exist will happen for the majority of projects.